node/core/approval-voting/src/lib.rs

```diff
@@ -723,13 +725,15 @@ enum Action {
 	Conclude,
 }
 
-async fn run<C>(
-	mut ctx: C,
+async fn run<Context>(
+	mut ctx: Context,
 	mut subsystem: ApprovalVotingSubsystem,
 	clock: Box<dyn Clock + Send + Sync>,
 	assignment_criteria: Box<dyn AssignmentCriteria + Send + Sync>,
 ) -> SubsystemResult<()>
-	where C: SubsystemContext<Message = ApprovalVotingMessage>
+where
+	Context: SubsystemContext<Message = ApprovalVotingMessage>,
+	Context: overseer::SubsystemContext<Message = ApprovalVotingMessage>,
```

Comment:
I don't quite get why this seemingly redundant trait bound is needed.

Reply:
It's this trade-off:

`Subsystem<Message=X> + overseer::Subsystem<Message=X>`

OR

`overseer::Subsystem<Message=X, AllMessages=AllMessages, Signal=OverseerSignal, Error=SubsystemError>`

since the other types need to be nailed down.

Reply:
I personally like the latter, because it's less magic even though it's more verbose. Is it possible to provide defaults fot the associated types except for the `Message`?

Reply:
They are unstable unfortunately, tried that as well.

Reply:
Weird, isn't one of those traits a super trait of the other?

Reply:
Is it possible to use generics with defaults for `AllMessages, Signal, Error` as a workaround?

Reply:
Yes, but that's not enough since `trait X: Y` and using `Context: X` does not resolve the `fn`s of `Y`.

Alt: Reversing the dependency `trait Y: X` and forwarding all `fn`s is impossible due to duplicate symbols in the final type and one has to use fully qualified syntax, which is an order of magnitude worse.

Reply:
> Is it possible to use generics with defaults for `AllMessages, Signal, Error` as a workaround?

You mean using smth. like `trait overseer::Subsystem<AllMessages=AllMessages, Signal=OverseerSignal, Error=SubsystemError> {  type Message; type Sender;  /* .. */}`.

Technically yes, but `SubsystemError` and the ident `AllMessages` are user provided. So that's generated code and at that point we're back to moving the traits into the proc-macro / `quote!{}` which was something I wanted to avoid

Reply:
But maybe for theose two traits it might be the better solution, in order to avoid all the clutter across all subsystems.

Reply:
I see :cry: . A simple `as_ctx()` method to expose the super trait would work, I guess. But not sure that's better.

Reply:
> > Is it possible to use generics with defaults for `AllMessages, Signal, Error` as a workaround?
> 
> You mean using smth. like `trait overseer::Subsystem<AllMessages=AllMessages, Signal=OverseerSignal, Error=SubsystemError> { type Message; type Sender; /* .. */}`.
> 
> Technically yes, but `SubsystemError` and the ident `AllMessages` are user provided. So that's generated code and at that point we're back to moving the traits into the proc-macro / `quote!{}` which was something I wanted to avoid

Yes, that what I was thinking. Maybe introduce yet another trait :grimacing:  close to user provided structs
```rust
trait RealSubsystemContext {
  type Message;
  ...
}
impl<T: RealSubsystemContext> for overseer::SubsystemContext {
  types AllMessages = AllMessages;
  ...
}
```

Reply:
It looks simple, but it's a rather evasive problem. The last thing that I can think of is providing a `Deref` impl for `Subsystem` decaying into `overseer::Subsystem` which could work.